File: LeetCode-In-Go/internal/dfs/LeetCode_110.go
package dfs

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}

	var (
		l int
		r int
	)
	if root.Left != nil {
		l = dfs(root.Left)
	}
	if root.Right != nil {
		r = dfs(root.Right)
	}
	if l-r > 1 || r-l > 1 {
		return false
	} else {
		return isBalanced(root.Left) && isBalanced(root.Right)
	}
}

func dfs(root *TreeNode) int {
	if root == nil {
		return 0
	}
	l := dfs(root.Left)
	r := dfs(root.Right)
	if l == 0 && r == 0 {
		return 1
	}
	return max(l, r) + 1
}

func max(r, l int) int {
	if r > l {
		return r
	}
	return l
}

File: LeetCode-In-Java/src/simple/depthFirstSearch/LeetCode_110.java
package simple.depthFirstSearch;

import lib.TreeNode;

/**
 * 应用场景：
 *
 * @author <a href="mailto:ge.sf.chn@gmail.com">shaofeng</a>
 * @since 2019/5/20
 */
public class LeetCode_110 {

    public Boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        int l = 0;
        int r = 0;
        if (root.left != null) {
            l = recusion(root.left);
        }
        if (root.right != null) {
            r = recusion(root.right);
        }
        if (l - r > 1 || r - l > 1) {
            return false;
        } else {
            return isBalanced(root.left) && isBalanced(root.right);
        }
    }

    public int recusion(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = recusion(root.left);
        int right = recusion(root.right);
        if (left == 0 && right == 0) {
            return 1;
        }
        return Math.max(left, right) + 1;
    }
}
